### PR TITLE
add method to filter unused classes

### PR DESCRIPTION
**Pull request by ghost**
this pull addresses issue https://github.com/capstone-coal/pycoal/issues/60 to add a `filterClasses` method to remove unused classes from a mineral-classified image and reindex the pixels accordingly. it also adds a test to verify behavior. the method is static because it does not depend on internal state. i expect a merge conflict with pull https://github.com/capstone-coal/pycoal/pull/66.

i tried it out on a number of images. it is much easier to see the classes in an image with QGIS, although some images still contain a confounding number of classes. it is now possible to see what classes an image contains just by looking at the header file.

it is now clear that some of the classifications are way off. what were classified as epsomite in the [classified west virginia image](https://drive.google.com/open?id=0BxysdOuBmaIGNmdiLUJ0a09ldEk) are clearly roads and lakes in the [rgb](https://drive.google.com/open?id=0BxysdOuBmaIGSHEzWHExSDJTRGs) version.

there are still too many classes in the [classified san juan image](https://drive.google.com/open?id=0BxysdOuBmaIGalNmVWhJRGVqUGM) to make things out clearly, although it is an improvement. the [histogram](https://drive.google.com/open?id=0BxysdOuBmaIGWFFMQkZIdmphdEk) is somewhat easier to interpret now. same story with one of the [high-res classified images](https://drive.google.com/open?id=0BxysdOuBmaIGMjl2TGwwY0N1T0E). i did find that the first of that batch, the one that took 70 hours, was classified incorrectly due to an unknown error.

i think this will be more powerful in combination with the [subset issue](https://github.com/capstone-coal/pycoal/issues/65) to let a user choose classes of interest.

i applied it automatically at the end of the mineral classifier to remove unused classes by default. i'm hoping this doesn't break anything. the mineral classification tests would help add confidence that this gives the correct behavior.

### CODE REVIEW

**Comment by ghost** [comment]
now that images don't necessarily contain all the classes in the spectral library, i had to add logic to the mining classifier to handle classes that are not present.

File: pycoal/mining.py [hunk 1]
```diff
@@ -51,7 +51,8 @@ def classifyImage(self, imageFilename, classifiedFilename):
         classified = numpy.zeros(shape=(M,N), dtype=numpy.uint16)
 
         # get class numbers from names
-        classNums = [image.metadata.get('class names').index(className) for className in self.classNames]
+        classList = image.metadata.get('class names')
+        classNums = [classList.index(className) if className in classList else -1 for className in self.classNames]
```
Comment (ghost) [inline pycoal/mining.py:55]: this could be unit tested with a filtered subimage

**Review by lewismc — APPROVED** [review]
+1

**Comment by ghost** [comment]
i've resolved the merge conflicts and am merging.

one improvement to this method would filter the most common classes from an image rather than all that are present, but that can be addressed in a separate issue. this would provide yet another way to slice and dice the classified data.

**Comment by ghost** [comment]
adding to the previous comment for future reference, filtering the "most common" classes could be interpreted several ways: finding the top `n` classes, classes with more than `m` instances or proportion `p` of the image, or classes above some statistical percentile of the data. as noted above, images may contain many distinct classes of which only a few are very common. this kind of filtering, unlike that which was just implemented, would throw away potentially useful data so it would be best not to overwrite the image.

**Pull request merged**